src/PrivateDns/PrivateDns/Models/PrivateDnsClient.cs

```diff
@@ -294,5 +294,292 @@ private static PSPrivateDnsVirtualNetworkLink ToPrivateDnsLink(VirtualNetworkLin
+
+        private RecordSet ConstructRecordSetProperties(
+            RecordType recordType,
+            uint? ttl,
+            Hashtable tags,
+            PSPrivateDnsRecordBase[] resourceRecords)
+        {
+
+            var properties = new RecordSet
+            {
+                Metadata = TagsConversionHelper.CreateTagDictionary(tags, validate: true),
+                Ttl = ttl
+            };
+
+            if (resourceRecords != null && resourceRecords.Length != 0)
+            {
+                var expectedTypeOfRecords = this.recordTypeValidationEntries[recordType];
+                var mismatchedRecord = resourceRecords.FirstOrDefault(x => x.GetType() != expectedTypeOfRecords);
+                if (mismatchedRecord != null)
+                {
+                    throw new ArgumentException(string.Format(ProjectResources.Error_AddRecordTypeMismatch, mismatchedRecord.GetType(), recordType));
+                }
+
+                FillRecordsForType(properties, recordType, resourceRecords);
+            }
+            else
+            {
+                FillEmptyRecordsForType(properties, recordType);
+            }
+
+            return properties;
+        }
+
+        private static void FillRecordsForType(RecordSet properties, RecordType recordType, IReadOnlyList<PSPrivateDnsRecordBase> resourceRecords)
+        {
+            switch (recordType)
+            {
+                case RecordType.A:
+                    properties.ARecords = resourceRecords.Select(x => (Sdk.ARecord)(x as ARecord)?.ToMamlRecord()).ToList();
+                    break;
+                case RecordType.AAAA:
+                    properties.AaaaRecords = resourceRecords.Select(x => (Sdk.AaaaRecord)(x as AaaaRecord)?.ToMamlRecord()).ToList();
+                    break;
+                case RecordType.CNAME:
+                    if (resourceRecords.Count > 1)
+                    {
+                        throw new ArgumentException(ProjectResources.Error_AddRecordMultipleCnames);
+                    }
+
+                    properties.CnameRecord = (Sdk.CnameRecord)resourceRecords[0].ToMamlRecord();
+                    break;
+                case RecordType.MX:
+                    properties.MxRecords = resourceRecords.Select(x => (Sdk.MxRecord)(x as MxRecord)?.ToMamlRecord()).ToList();
+                    break;
+                case RecordType.PTR:
+                    properties.PtrRecords = resourceRecords.Select(x => (Sdk.PtrRecord)(x as PtrRecord)?.ToMamlRecord()).ToList();
+                    break;
+                case RecordType.SRV:
+                    properties.SrvRecords = resourceRecords.Select(x => (Sdk.SrvRecord)(x as SrvRecord)?.ToMamlRecord()).ToList();
+                    break;
+                case RecordType.TXT:
+                    properties.TxtRecords = resourceRecords.Select(x => (Sdk.TxtRecord)(x as TxtRecord)?.ToMamlRecord()).ToList();
+                    break;
+                case RecordType.SOA:
```

Comment:
How come SOA does not have the same validation as CNAME (L365-368)?

Reply:
is there a secnario where we can have multiple records in a SOA recordset? afaik manual creation is disallowed in soa records. so why even have that check?